### ISSUE TITLE
[FEATURE REQ] - Make Azure.Messaging.EventHubs.EventProcessorClient.ProcessEventAsync virtual so that it can be mocked and raised in a unit test scenario.

### ISSUE THREAD

**Issue by judebowman** [Azure/azure-sdk-for-net#42868]
### Library name

Azure.Messaging.EventHubs

### Please describe the feature.

Very much related to the question in https://github.com/Azure/azure-sdk-for-net/issues/31533

The feature request is to make Azure.Messaging.EventHubs.EventProcessorClient.ProcessEventAsync virtual so that it can be mocked and raised in a unit test scenario.  

You can see a sample repo at https://github.com/denisivan0v/event-processor-client-integration-tests/blob/main/SampleApp.Tests/IntegrationTests.cs with the class under test living at https://github.com/denisivan0v/event-processor-client-integration-tests/blob/main/SampleApp/Worker.cs

In particular, at line 54 of the ltegrationTests class, you can see we can not use Moq to raise the event because it is not virtual.

The proposed solutions to the above linked query issue both depend on and assume that the wrapper class make their ProcessEventAsync handlers public.  But in both the linked repo (in https://github.com/denisivan0v/event-processor-client-integration-tests/blob/main/SampleApp/Worker.cs), and the wrapper class I am currently working on in a enterprise code base, these event handler and error handler methods are private, and I am not a big fan of making them public just for unit testing purposes.

CC: @m-redding and @jsquire and @denisivan0v

**Comment by judebowman** [comment]
The sample repo was provided by @denisivan0v, and I am very glad that I stumbled upon their initial query issue https://github.com/Azure/azure-sdk-for-net/issues/31533 because it is the exact scenario I am dealing with right now.

**Comment by judebowman** [comment]
I should include that if it is decided to move forward with this request to make the class more unit testable, the same should probably be done for the other non-virtual events, such as ProcessErrorAsync, and perhaps even PartitionInitializingAsync and PartitionClosingAsync.

**Comment by judebowman** [comment]
I do not think my plan to use a fake which inherits from the real EventProcessorClient will work now, because StartProcessingAsync is not virtual and thus cannot be overridden.   :weary:

**Comment by judebowman** [comment]
Now that I have implemented the wrapper around EventProcessorClient to make mocking very easy, I think this is probably the way to go.  I think I understand why the actual SDK does not utilize interfaces around the classes to make mocking easier - backwards compatibility and breaking changes when adding new public method to the classes, where doing so to a class is not a breaking change, but when using an interface, then the adding of public methods/properties/events to the interface would be a breaking change.

Anyhow, here is what my current implementation to get around this issue looks like:

`
public interface IEventProcessorClient
{
    Task StartProcessingAsync(CancellationToken cancellationToken = default);
    Task StopProcessingAsync(CancellationToken cancellationToken = default);

    event Func<ProcessEventArgs, Task> ProcessEventAsync;
    event Func<ProcessErrorEventArgs, Task> ProcessErrorAsync;
}

public class EventProcessorClientWrapper : IEventProcessorClient
{
    private readonly EventProcessorClient _eventProcessorClient;

    public EventProcessorClientWrapper(BlobContainerClient checkpointStore, string consumerGroupName, string fullyQualifiedNamespace, string eventHubName, TokenCredential tokenCredential)
    {
        _eventProcessorClient = new EventProcessorClient(checkpointStore, consumerGroupName, fullyQualifiedNamespace, eventHubName, tokenCredential);
    }

    public Task StartProcessingAsync(CancellationToken cancellationToken = default)
    {
        return _eventProcessorClient.StartProcessingAsync(cancellationToken);
    }

    public Task StopProcessingAsync(CancellationToken cancellationToken = default)
    {
        return _eventProcessorClient.StopProcessingAsync(cancellationToken);
    }

    public event Func<ProcessEventArgs, Task> ProcessEventAsync
    {
        add => _eventProcessorClient.ProcessEventAsync += value;
        remove => _eventProcessorClient.ProcessEventAsync -= value;
    }

    public event Func<ProcessErrorEventArgs, Task> ProcessErrorAsync
    {
        add => _eventProcessorClient.ProcessErrorAsync += value;
        remove => _eventProcessorClient.ProcessErrorAsync -= value;
    }
}
`

**Comment by m-redding** [comment]
Hello @judebowman! Thank you for reaching out with your question & feature request. Unfortunately, we won't be able to move forward with this as a feature, but I am happy to help answer questions about mocking. I had a couple clarifications and questions:

> Further, I noticed that StartProcessingAsync and StopProcessingAsync are not virtual (likely because they are already themselves overrides of virtual methods in the EventProcessorClient's base class).

I just wanted to clarify that `StartProcessingAsync` and `StopProcessingAsync` are virtual, which should be helpful.

> the wrapper class I am currently working on in a enterprise code base, these event handler and error handler methods are private, and I am not a big fan of making them public just for unit testing purposes.

Could you make your handlers internal? You linked the other issue about mocking, there I mentioned that in general we recommend focusing tests on handlers, since we have an extensive test suite to validate the behavior of the `EventProcessorClient`. How are you testing these handlers now?

The other issue was focused on adding an integration test, so we added this sample to help cover that case: [Simulating a running processor](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/Sample08_MockingClientTypes.md#simulating-the-eventprocessorclient-running). I'm hoping this will be helpful if you're looking to test a full end to end scenario.

I did look at the attached screenshots and I don't totally follow the exact scenario you're hoping to test. You can assume the handlers will get called at the right times. Would you mind giving me a little more information about the scenario you're trying to test? That way I can give some more specific guidance.

**Comment by judebowman** [comment]
Thanks for the response @m-redding!  Let me put together a small repo which shows what my base class is attempting to do, and the unit tests I ended up writing for it, to help you better understand.  I will comment on here shortly once it is ready.

**Comment by judebowman** [comment]
Hi @m-redding, I have added a repository which is similar to what I am trying to do here - https://github.com/judebowman/event-hubs

The main class in question is the [EventStreamConsumer](https://github.com/judebowman/event-hubs/blob/main/Foundation.EventStreaming.EventHubs/Consumer/EventStreamConsumer.cs).

It's main responsibility is to act as a base class for many different consumer group instances hosted as background services on different on-premise servers for various event types.  It takes care of a few things we did not want each consumer application to have to duplicate, including:

-the handling of saving checkpoint updates to the blob storage on a configurable interval with a reasonable default
- a configurable number of automatic, inline retries with an exponential back off which are meant to help with transient errors which may be resolved with an amount of time between retries.
- poison events which can not be deserialized to their intended generic type.
- delegating the event handler for receiving a message which wraps the event data in a generic event envelope so that our consumers don't have to directly have to depend on the Azure ProcessEventArgs type for the happy path.
- providing an "OnFinalException" event handler which is triggered when all of the retries have resulted in an exception.  The idea is to give the consumer implementation one last chance to save the details of the failed event in a poison message store (likely a database table) and triggering notifications so that the issue can be looked into, and anything else application specific they need to do in the event of not being able to successfully process the event.
- Application Insights logging 
- Other things but those are the main responsibilities

In my current design, I have no reason to make `EventStreamConsumer.ProcessEventAsync()` public.  It is meant to be private, and thus not unit testable directly.  Specifically, from the perspective of the classes which inherit from this base class, they only need to know about the following public interface:

![image](https://github.com/Azure/azure-sdk-for-net/assets/1609183/c803a79f-e045-40ee-a2b4-05487b1f9729)

So, in order to test that my base class does all the things it needs to do when it receives an event from the Azure      EventProcessorClient, I need to call the method `EventStreamConsumer.StartProcessingAsync()` and then trigger my dependency (the event processor client) to raise the event for when an event is ready to process.

With my [EventProcessorClientWrapper ](https://github.com/judebowman/event-hubs/blob/main/Foundation.EventStreaming.EventHubs/Consumer/EventProcessorClientWrapper.cs) class, I am able to mock the raising of the event [here](https://github.com/judebowman/event-hubs/blob/85ff8f2eda941344fa10e0c3e7f337917df5ead0/Foundation.EventStreaming.EventHubs.Tests/Consumer/EventStreamConsumerTests.cs#L422)

**Comment by judebowman** [comment]
I know my unit testing nested classes are unconventional, but when you look at them in the ReSharper test runner, I think they tell a good story of what test cases and expected outcomes I am trying to codify:

![image](https://github.com/Azure/azure-sdk-for-net/assets/1609183/cc574efa-09a5-4c97-b488-c4eb1c5ca03c)

I think I did a good job of covering nearly all of the logic and different possible inputs and outputs that can happen.

Unfortunately, I was not able to achieve this ability to unit test in the same way with using a mock of `EventProcessorClient`, because the event `ProcessEventAsync()` is not virtual and thus not mockable.

I know I just threw a lot at you, but I figured it was better to share more than less.  Please let me know if what I shared makes sense or not, and if you have any clarifying or follow up questions!

**Comment by m-redding** [comment]
Hello @judebowman thank you for the additional information! I'm still a bit confused by this statement here:
> In my current design, I have no reason to make [EventStreamConsumer.ProcessEventAsync()](https://github.com/judebowman/event-hubs/blob/85ff8f2eda941344fa10e0c3e7f337917df5ead0/Foundation.EventStreaming.EventHubs/Consumer/EventStreamConsumer.cs#L148) public. It is meant to be private, and thus not unit testable directly.

Is there a reason why they cannot be `internal`? I see that you are already making the internals visible to your test project [here](https://github.com/judebowman/event-hubs/blob/main/Foundation.EventStreaming.EventHubs/Foundation.EventStreaming.EventHubs.csproj#L16). I totally understand not wanting to make them public - but I'm missing why they cannot be internal?

`internal async Task ProcessEventAsync(ProcessEventArgs processEventArgs)`
`internal async Task ProcessErrorAsync(ProcessErrorEventArgs processErrorEventArgs)`

Then you could tweak your tests to be something like the following:
```csharp
public class When_events_are_received
{
    public class With_no_exceptions_thrown : TestBase
    {
        private readonly EventStreamConsumer<TestEventType> _consumer;
        private readonly ProcessEventArgs _processEventArgs;
        private int _numberOfOnEventReceivedInvocations;
        private int _numberOfOnFinalExceptionInvocations;
        private int _numberOfUpdateCheckpointInvocations;
        private readonly TelemetryChannelFake _telemetryChannelFake;

        public With_no_exceptions_thrown()
        {
            _telemetryChannelFake = new TelemetryChannelFake();
            var telemetryClient = GetTelemetryClient(_telemetryChannelFake);

            _consumer = CreateConsumer<TestEventType>(AutoMocker, telemetryClient);
            _consumer.OnEventReceived += OnEventReceived;
            _consumer.OnFinalException += OnFinalException;
            _consumer.OnEventStreamFailure += _ => Task.CompletedTask;
            
            _processEventArgs = GetProcessEventArgs(UpdateCheckpointCallback);

            SetupMocks();
        }

        private void SetupMocks()
        {
            SetupServiceScope();
        }

        private Task UpdateCheckpointCallback(CancellationToken cancellationToken)
        {
            _numberOfUpdateCheckpointInvocations++;
            return Task.CompletedTask;
        }

        private Task OnEventReceived(EventStreamData<TestEventType> eventStreamData, IServiceScope serviceScope)
        {
            _numberOfOnEventReceivedInvocations++;
            return Task.CompletedTask;
        }

        private Task OnFinalException(Exception arg1, ProcessEventArgs arg2, IServiceScope arg3)
        {
            _numberOfOnFinalExceptionInvocations++;
            return Task.CompletedTask;
        }

        [Fact]
        public async Task Then_OnEventReceived_should_be_invoked_once()
        {
            var cancellationTokenSource = new CancellationTokenSource();
            //await _consumer.StartProcessingAsync(cancellationTokenSource);

            //await AutoMocker.GetMock<IEventProcessorClient>().RaiseAsync(x => x.ProcessEventAsync += null, _processEventArgs);

            await _consumer.ProcessEventAsync(_processEventArgs);

            _numberOfOnEventReceivedInvocations.Should().Be(1);
        }

        [Fact]
        public async Task Then_OnFinalException_should_not_be_invoked()
        {
            var cancellationTokenSource = new CancellationTokenSource();
            //await _consumer.StartProcessingAsync(cancellationTokenSource);

            //await AutoMocker.GetMock<IEventProcessorClient>().RaiseAsync(x => x.ProcessEventAsync += null, _processEventArgs);

            await _consumer.ProcessEventAsync(_processEventArgs);

            _numberOfOnFinalExceptionInvocations.Should().Be(0);
        }

        [Fact]
        public async Task Then_UpdateCheckpointAsync_should_be_invoked_after_the_configured_wait_time()
        {
            var cancellationTokenSource = new CancellationTokenSource();
            //await _consumer.StartProcessingAsync(cancellationTokenSource);

            //await AutoMocker.GetMock<IEventProcessorClient>().RaiseAsync(x => x.ProcessEventAsync += null, _processEventArgs);
            
            await _consumer.ProcessEventAsync(_processEventArgs);

            await Task.Delay(1000, default);

            _numberOfUpdateCheckpointInvocations.Should().Be(1);
        }
```
You may need some additional changes to make mocking easier, such as adding an internal constructor, so you can set mock objects for things like [this](https://github.com/judebowman/event-hubs/blob/main/Foundation.EventStreaming.EventHubs/Consumer/EventStreamConsumer.cs#L33)

**Comment by judebowman** [comment]
@m-redding, I agree that what you propose is possible, and will allow me to directly unit test the method that went from private to internal, but I feel like changing the access level of a method to accomplish unit testing is a code smell and somewhat of a work around/cop-out -- not very much different than making a private method public just to unit test it directly instead of unit testing the public method that calls the private method.

Also, if I went with the proposed technique, I would not have test coverage for the relationship between my `EventStreamConsumer` class and it's dependency of the `EventProcessorClient` -- specifically around the logic that when `EventProcessorClient` raises an event, `EventStreamConsumer` has a handler wired to the event and that event handler does things in response to the event being raised.

Specifically, I would lose coverage around this logic:

<pre>
     _eventProcessorClient.ProcessEventAsync += ProcessEventAsync;
     _eventProcessorClient.ProcessErrorAsync += ProcessErrorAsync;
</pre>

In my current version of the unit tests, if someone where to remove those two lines, the tests would fail, thus giving valuable visibility into the need for the above lines of code and causing a PR to fail until the issue is resolved.

But if I went with the proposed suggestion, and someone removed the above two lines of code, the tests would still pass and we wouldn't know there is an issue until the code is deployed to an actual environment. 

I hope that makes sense, but let me know if you have any follow up questions.

I do appreciate you taking the time to learn more about my issue and your follow ups.  I know you said early on that you all will not be able to move forward with my request, and I understand that.  I am curious if you would be able to share the reason behind that decision, if you are able to share.

I will stick with my my [EventProcessorClientWrapper ](https://github.com/judebowman/event-hubs/blob/main/Foundation.EventStreaming.EventHubs/Consumer/EventProcessorClientWrapper.cs) class and interface for now since it meets my needs of ensuring all the logic I want to verify is able to be captured in unit tests, and as it also provides me some flexibility and decoupling.   Although if I were able to mock the `EventProcessorClient.ProcessEventAsync` event (by it being virtual), I would have preferred to do so.

Also, I hope my real world code example shows some of the interesting things that are being done with the SDK's `EventProcessorClient`, and also some of the base functionality that we felt like we needed to add to it to be used in many different scenarios.

Thanks again!

**Comment by m-redding** [comment]
@judebowman 
> I am curious if you would be able to share the reason behind that decision, if you are able to share.

Yes! It basically just comes down to virtual events not being a standard usage pattern in .NET. It is possible to declare virtual events, and I'm sure there are use cases out there, but they can't be overridden properly, some info about this [here](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/events/how-to-raise-base-class-events-in-derived-classes). Instead, the standard pattern is to have `On<Event>` methods that can be overridden, like [this](https://learn.microsoft.com/en-us/dotnet/standard/events/#events). In general, the Azure SDKs prioritize being idiomatic to their respective languages even if the existing patterns aren't perfect.

> I will stick with my my [EventProcessorClientWrapper ](https://github.com/judebowman/event-hubs/blob/main/Foundation.EventStreaming.EventHubs/Consumer/EventProcessorClientWrapper.cs)class and interface for now since it meets my needs of ensuring all the logic I want to verify is able to be captured in unit tests, and as it also provides me some flexibility and decoupling.

I'm glad that you were able to find a suitable solution even though it's not the most ideal or elegant. If you were interested in some ideas around covering the handler integration, I could provide some but given what you mentioned here I'm thinking you have what you need for your testing.

I'm going to close this issue as resolved, but feel free to open a new issue if any other questions/requests come up!
